Add enable/disable api for SystemRuntime plugins

dmpots opened the pull request:
This commit adds support for enabling and disabling plugins by name. The changes are made generically in the `PluginInstances` class, but currently we only expose the ability to SystemRuntime plugins. Other plugins types can be added easily.

We had a few design goals for how disabled plugins should work

  1. Plugins that are disabled should still be visible to the system. This allows us to dynamically enable and disable plugins and report their state to the user.
  2. Plugin order should be stable across disable and enable changes. We want avoid changing the order of plugin lookup. When a plugin is re-enabled it should return to its original slot in the creation order.
  3. Disabled plugins should not appear in PluginManager operations. Clients should be able to assume that only enabled plugins will be returned from the PluginManager.

For the implementation we modify the plugin instance to maintain a bool of its enabled state.  Existing clients external to the Instances class expect to iterate over only enabled instance so we skip over disabed instances in the query and snapshot apis. This way the client does not have to manually check which instances are enabled.

dmpots committed:
Add enable/disable api for SystemRuntime plugins

This commit adds support for enabling and disabling plugins by name.
The changes are made generically in the `PluginInstances` class, but
currently we only expose the ability to SystemRuntime plugins. Other
plugins types can be added easily.

We had a few design goals for how disabled plugins should work

  1. Plugins that are disabled should still be visible to the system.
     This allows us to dynamically enable and disable plugins
     and report their state to the user.
  2. Plugin order should be stable across disable and enable changes.
     We want avoid changing the order of plugin lookup. When a plugin
     is re-enabled it should return to its original slot in the
     creation order.
  3. Disabled plugins should not appear in PluginManager operations.
     Clients should be able to assume that only enabled plugins will
     be returned from the PluginManager.

For the implementation we modify the plugin instance to maintain a bool
of its enabled state.  Existing clients external to the Instances class
expect to iterate over only enabled instance so we skip over disabed
instances in the query and snapshot apis. This way the client does not
have to manually check which instances are enabled.

dmpots commented:
This PR modifies the PluginManager to support enable/disable. A following [commit](https://github.com/dmpots/llvm-project/commit/aaa3ab3d063580747bb106f39e6792232cc0af00) will expose the commands to the user.

This is currently only working for the SystemRuntime plugins. Once these first commits land I plan to do the same for the other plugin types.

JDevlieghere reviewed (approved):
This change LGTM. 

Are you able to share the motivation behind wanting to enable/disable plugins at runtime? 

I added support for disabling plugins at configuration/build time and at the time @labath had concerns about that as various parts of LLDB depend on certain plugins. Do you plan on making this opt-in for plugins or will users be able to disable arbitrary plugins at runtime?

dmpots commented:
> Are you able to share the motivation behind wanting to enable/disable plugins at runtime?

The motivation is to enable easier testing and experimentation.

It came from our attempts to try lazy symbol loading (i.e. `settings set target.preload-symbols false`).  We found that even with preload-symbols disabled there were various things that would trigger symbol loading unconditionally. 

We fixed a source of symbol loading in #129593, but found that there were several plugins that would force load all symbols. Some of these plugins (e.g. ASAN) are loaded unconditionally with no way to disable them. This not only makes it hard to test, but also harder to do one-off experiments without modifying the code and rebuilding.

After discussing with @clayborg we thought a generic mechanism to disable plugins would be useful and could help with things beyond our original motivation.

We also have some internal patches to disable some of these plugins, but if we have a generic way to do it then we can add it to our custom lldb init. This should reduce the number of internal diffs we have.

One last idea I should mention is that we thought we could also extend this mechanism to support reordering of plugins. So that, for example, when running on linux we could reorder the ObjectFile plugins to have the Elf plugin come first. Just an idea at this point, but something we might explore in the future.

> I added support for disabling plugins at configuration/build time and at the time @labath had concerns about that as various parts of LLDB depend on certain plugins. Do you plan on making this opt-in for plugins or will users be able to disable arbitrary plugins at runtime?

The plan is to allow disabling arbitrary plugins at runtime. We could have an opt-in or opt-out mechanism, but my feeling is that giving the user full control is not too bad. Most end users will not touch it and lldb developers can manage their self-imposed difficulties.

dmpots commented:
> LLVM Buildbot has detected a new failure on builder `lldb-aarch64-ubuntu` running on `linaro-lldb-aarch64-ubuntu` while building `lldb` at step 6 "test".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/59/builds/15347
> 
> Here is the relevant piece of the build log for the reference

I suspect this is related to #134048, which is also seeing the failure.

dmpots commented:
I have an attempted fix for these failures in #134173. I suspect ICF is combining the create functions into a single address which make us unable to remove them based on a unique address (which the plugin manager assumes).
